apps/events/migrations/0011_modify_schedulemessage_class.py

```diff
@@ -0,0 +1,63 @@
+# Generated by Django 4.2.11 on 2024-06-19 17:15
+from datetime import datetime, timedelta
+from django.db import migrations
+
+from apps.events.models import EventAction, EventActionType, ScheduledMessage
+from apps.experiments.models import ExperimentSession
+
+def is_complete(task):
+	return datetime.now() > task.expires
+
+def calculate_next_run(task, current_time, is_complete):
+	if is_complete:
+		return None
+	delta = timedelta(seconds=task.interval.every.total_seconds())
+	next_time = task.start_time
+	while next_time <= current_time:
+		next_time += delta
+	return next_time
+
+def calculate_last_run(task, current_time, is_complete):
+	if is_complete:
+		return task.expires
+	delta = timedelta(seconds=task.interval.every.total_seconds())
+	last_time = task.start_time
+	while last_time <= datetime.now():
+		last_time += delta
+	return last_time - delta
+
+
+class Migration(migrations.Migration):
+    def migrate_periodic_tasks(apps, schema_editor):
+        PeriodicTask = apps.get_model('django_celery_beat', 'PeriodicTask')
+        for periodic_task in PeriodicTask.objects.filter(name__startswith="reminder-"):
+            parts = periodic_task.name.split("-")
+            if len(parts) >= 3:
+                experiment_session_id = parts[1]
+                experiment_session = ExperimentSession.objects.get(id=experiment_session_id)
+                params = {
+                    "time_period": periodic_task.kwargs.get("time_period", "day"),
+                    "frequency": periodic_task.kwargs.get("frequency", 1),
+                    "repetitions": periodic_task.kwargs.get("repetitions", 1),
```

Comment:
looks like `periodic_task.kwargs` returns a json. I think the following should help

```python
from apps.events.models import EventActionType
import json

def calc_schedule_components(periodic_task):
    """Periodic task can be a one-off or interval task.
    
    one-off uses the clocked schedule
    interval uses the interval schedule
    """
    if periodic_task.one_off:
        schedule = periodic_task.clocked
        frequency = 1
        # PeriodicTask doesn't have a created at type of field, making it impossible to determine the time_period e.g. hours, days, weeks etc..let's keep it null. This is a doozy. Either we keep it None or assume some time period. I'm leaning to having it None to make it fail if the schedule is updated, which I doubt will happen tbh
        time_period = None
        repetitions = 1
        next_trigger_date = schedule.clocked_time
        total_triggers = 0
    else:
        schedule = periodic_task.interval
        frequency = schedule.every
        time_period = schedule.period
        repetitions = ... # needs to be calculated using expires
        next_trigger_date = ... # needs to be calculated using calculate_next_run
        total_triggers = ... # needs to be calculated

    return frequency, time_period, repetitions, total_triggers, next_trigger_date 

def create_scheduled_messages(apps, schema_editor):
    ScheduledMessage = apps.get_model('events', 'ScheduledMessage')
    EventAction = apps.get_model('events', 'EventAction')
    Experiment = apps.get_model('experiments', 'Experiment')
    Participant = apps.get_model('experiments', 'Participant')

    task = ... # filter for active ones
    # filter for tasks that are disabled
    task_kwargs = json.loads(task.kwargs)
    # There may be many participants, so we have to create a scheduled message + event for each
    participant_identifiers = task_kwargs["chat_ids"]
    message = task_kwargs["message"]
    experiment = Experiment.objects.filter(public_id=task_kwargs["experiment_public_id"])

    frequency, time_period, repetitions, total_triggers, next_trigger_date = calc_schedule_components(task)

    event_action_params = {
        "name": "", # we can generate some name
        "prompt_text": message,
        "frequency": frequency,
        "time_period": time_period,
        "repetitions": repetitions,
        "experiment_id": experiment.id
    }

    for identifier in participant_identifiers:
        # There may be many participants with the same identifier in other teams, so we need to filter by team as well
        participant = Participant.objects.get(team=experiment.team, identifier=identifier)

        event_action = EventAction.objects.create(
            action_type=EventActionType.SCHEDULETRIGGER,
            params=event_action_params
        )
        ScheduledMessage.objects.create(
            experiment=experiment,
            participant=participant,
            team=experiment.team,
            action=event_action,
            last_triggered_at=task.last_run_at,
            next_trigger_date=next_trigger_date,
            total_triggers=total_triggers
        )
```